### PR TITLE
578 components connectness origin

### PR DESCRIPTION
**Pull request by akokoshn**
CI: https://github.com/NilFoundation/zkLLVM/pull/596

### CODE REVIEW

File: include/nil/blueprint/recursive_prover/fri_array_swap.hpp [hunk 1]
```diff
@@ -75,15 +75,13 @@ namespace nil {
             std::vector<var> res = get_component_result<BlueprintFieldType, component_type>
                     (bp, assignment, internal_storage, statistics, param, instance_input, array_size / 2).output;
 
-            if (param.gen_mode.has_assignments()) {
                 ptr_type result_ptr = static_cast<ptr_type>(typename BlueprintFieldType::integral_type(
```
Comment (makxenov) [inline include/nil/blueprint/recursive_prover/fri_array_swap.hpp:78]: The same about indentation

File: include/nil/blueprint/recursive_prover/fri_cosets.hpp [hunk 2]
```diff
@@ -68,7 +68,6 @@ namespace nil {
             const auto& result = get_component_result<BlueprintFieldType, component_type>
                 (bp, assignment, internal_storage, statistics, param, instance_input, res_length, omega).output;
 
-            if (param.gen_mode.has_assignments()) {
                 ptr_type result_ptr = static_cast<ptr_type>(
```
Comment (makxenov) [inline include/nil/blueprint/recursive_prover/fri_cosets.hpp:71]: The same

File: include/nil/blueprint/assigner.hpp [hunk 3]
```diff
@@ -946,11 +967,58 @@ namespace nil {
...
                 gen_mode = assigner_state.gen_mode;
                 finished = assigner_state.finished;
+                memory.restore_state(assigner_state.mem_state);
+            }
+
+            void merge_memory_state(const memory_state<var>& state, const var& cond) {
+                auto stack_top = std::max(memory.get_stack_top(), state.stack_top);
+                auto heap_top = std::max(memory.get_heap_top(), state.heap_top);
+                const common_component_parameters param = {targetProverIdx, gen_mode};
+                auto merge_region = [&cond, &param, &state, this](size_t begin, size_t end, size_t other_end) {
```
[the remark this answers is not in the document]
Comment (akokoshn) [reply include/nil/blueprint/assigner.hpp:977]: Actually it's not about branch, but about memory merge region (stack or heap). So what do you think about `memory_region_begin`, `memory_region_end`, (current memory) `other_memory_region_and` (which should be merged with current)?

[the remark this answers is not in the document]
Comment (makxenov) [reply include/nil/blueprint/assigner.hpp:977]: Well, I particularly don't like `other` word. It kinda makes both ends symmetrical, but you process them differently. So I would like that the name will specify the fact that we must move values from this region and keep values from the other one.

File: include/nil/blueprint/assigner.hpp [hunk 4]
```diff
@@ -540,23 +556,19 @@ namespace nil {
...
                             ptr_type dst = resolve_number<ptr_type>(frame, inst->getOperand(0));
                             ptr_type src = resolve_number<ptr_type>(frame, src_val);
                             unsigned width = resolve_number<unsigned>(frame, inst->getOperand(2));
                             memcpy(dst, src, width);
-                        }
                         return true;
                     }
                     case llvm::Intrinsic::memset: {
-                        if (gen_mode.has_assignments()) {
                             ptr_type dst = resolve_number<ptr_type>(frame, inst->getOperand(0));
```
[the remark this answers is not in the document]
Comment (akokoshn) [reply include/nil/blueprint/assigner.hpp]: done

File: include/nil/blueprint/assigner.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (akokoshn) [reply include/nil/blueprint/assigner.hpp]: i just inspired by usual naming in copy constructors `class_name(class_name &other)`, but ok with another name, like `true_memory_region_end`, or what is your suggestion?

File: include/nil/blueprint/assigner.hpp [hunk 5]
```diff
@@ -946,11 +967,58 @@ namespace nil {
...
+            void merge_memory_state(const memory_state<var>& state, const var& cond) {
+                auto stack_top = std::max(memory.get_stack_top(), state.stack_top);
+                auto heap_top = std::max(memory.get_heap_top(), state.heap_top);
+                const common_component_parameters param = {targetProverIdx, gen_mode};
+                auto merge_region = [&cond, &param, &state, this](size_t begin, size_t end, size_t other_end) {
+                    auto max_end = std::max(end, other_end); // max memory state and current memory used cells
+                    // run throw all cells
+                    for (size_t i = begin; i < max_end; i++) {
+                        if (i < end && i < other_end) {
+                            auto v_true = (begin == 1) ? state.stack[i - begin].v : state.heap[i - begin].v;
```
[the remark this answers is not in the document]
Comment (akokoshn) [reply include/nil/blueprint/assigner.hpp]: done

File: include/nil/blueprint/recursive_prover/fri_array_swap.hpp (hunk 1 above)
Comment (akokoshn) [reply include/nil/blueprint/recursive_prover/fri_array_swap.hpp]: done

File: include/nil/blueprint/recursive_prover/fri_cosets.hpp (hunk 2 above)
Comment (akokoshn) [reply include/nil/blueprint/recursive_prover/fri_cosets.hpp]: done

File: include/nil/blueprint/assigner.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (makxenov) [reply include/nil/blueprint/assigner.hpp]: It looks good, thanks

**Review by aleasims — APPROVED** [review]

File: include/nil/blueprint/assigner.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (akokoshn) [reply include/nil/blueprint/assigner.hpp]: done

**Pull request merged**